scripts/xtensa-build-zephyr.py

```diff
@@ -0,0 +1,390 @@
+		input_elf_file = pathlib.Path(west_top, platform_build_dir_name, "zephyr", "zephyr.elf")
+		# Extract metadata
+		execute_command([str(smex_executable), "-l", str(fw_ldc_file), str(input_elf_file)],
+			check=True)
+		# Update SOF submodules
+		git_submodules_update()
+		# CMake - configure rimage module
+		rimage_dir_name="build-rimage"
+		sof_mirror_dir = pathlib.Path("modules", "audio", "sof")
+		rimage_source_dir = pathlib.Path(sof_mirror_dir, "rimage")
+		execute_command(["cmake", "-B", rimage_dir_name, "-S", str(rimage_source_dir)], check=True,
+			cwd=west_top)
+		# CMake build rimage module
+		execute_command(["cmake", "--build", rimage_dir_name, "-j", args.jobs], check=True,
+			cwd=west_top)
+		# Sign firmware
+		rimage_executable = shutil.which("rimage", path=pathlib.Path(west_top, rimage_dir_name))
+		rimage_config = pathlib.Path(sof_mirror_dir, "rimage", "config")
+		sign_cmd = ["west", "sign", "--build-dir", platform_build_dir_name, "--tool", "rimage"]
+		sign_cmd += ["--tool-path", rimage_executable]
+		signing_key = ""
+		if args.key:
+			signing_key = args.key
+		elif "RIMAGE_KEY" in platform_dict:
+			signing_key = platform_dict["RIMAGE_KEY"]
+		else:
+			signing_key = default_rimage_key
+		sign_cmd += ["--tool-data", str(rimage_config), "--", "-k", signing_key]
+		execute_command(sign_cmd, check=True, cwd=west_top)
+		# Install by copy
+		fw_file_to_copy = pathlib.Path(west_top, platform_build_dir_name, "zephyr", "zephyr.ri")
+		fw_file_installed = pathlib.Path(STAGING_DIR, "community", f"sof-{platform}.ri")
+		shutil.copy2(str(fw_file_to_copy), str(fw_file_installed))
+
+	# Install sof-logger
+	sof_logger_executable_to_copy = pathlib.Path(west_top, platform_build_dir_name, "zephyr",
+		"sof-logger_ep", "build", "logger", "sof-logger")
+	sof_logger_installed = pathlib.Path(STAGING_DIR, "..", "tools", "sof-logger").resolve()
+	shutil.copy2(str(sof_logger_executable_to_copy), str(sof_logger_installed))
+
+def run_clone_mode():
+	if find_west_workspace():
+		raise RuntimeError("Zephyr found already! Not downloading it again")
+	create_zephyr_directory()
+	west_init_update()
+	git_submodules_update()
+	build_platforms()
+
+def run_link_mode():
+	global west_top, SOF_TOP
+	west_workspace_path = find_west_workspace()
+	if not west_workspace_path:
+		raise RuntimeError("Failed to find west workspace. Scanned directories:\n{}\n{}\n{}"
+			.format(os.getcwd(), SOF_TOP, west_top))
+	if not west_workspace_path.exists():
+		raise FileNotFoundError("West topdir returned {} as workspace but it"
+			" does not exist".format(west_workspace_path))
+	create_zephyr_sof_symlink()
+	git_submodules_update()
+	build_platforms()
+
+def main():
+	parse_args()
+	check_west_installation()
+	if len(args.platforms) == 0:
+		print("No platform build requested")
+	else:
+		print("Building platforms: {}".format(" ".join(args.platforms)))
+
+	# Two mutually exclusive working modes are supported
```

Comment:
Cloning and linking are not mutually exclusive. The two options that are mutually exclusive are:
- Pointing at an _existing_ zephyr clone with `-p`
- Downloading a _new_ zephyr clone with `-c`

These are not two "modes", they're just two (obviously) mutually exclusive command line flags and features.

For now just copy the logic from the `main()` function in the original shell script, it's only 50 lines of commented code. It's also well tested so it barely needs testing again. You almost don't even need to understand it :-) Copy the comments too.

Reply:
@marc-hb I guess we have some misunderstanding here.
I named it "link_mode" as it "links existing zephyr directory using symlink" that's what I meant (not the linking process of linker). Logic is copied as is but with some additional features. That being said, that .sh script is not a holy grail and I can and will reword, reinterpret or rephrase content of it as a developer of this script as a part of my contribution.
Those two arguments "-c" and "-p" are not "flags". A flag provides user input to some selected program behavior. Those are "features" but not in the context of this code and review. They are exactly what i called them -" modes". Two different program logic flows. Can we please accept my point of view and understanding? I think they are perfectly valid and should not be a subject of this review. We should focus on finding bugs/errors in the script or use cases when it does not do what it supposed to do.

Reply:
Changed function name to run_point_mode() for consistency.

Reply:
> Can we please accept my point of view and understanding for now?

"flags", "modes", etc. I don't really care how you call them; afraid you lost me in the various definitions and nuances. My point was entirely the first sentence: "Cloning and linking are not mutually exclusive", I think that's why nothing built the first time I tried to use this script. I thought the "modes" terminology would help explain the problem but forget it, let's leave terminology aside.

I don't understand why you have to "fight argparser", some argument validation and processing can done _after_ invoking argparse, it does not have to be entirely implemented in argparse.

Last but not least, do not "resolve" and hide discussions after asking  "can you accept my point of view?" + leaving the "mutually exclusive" comment in place. Discussions are "resolved" only after agreement and it's fine to leave some discussions never resolved / never hidden - and easier to see even after merge.